File: algorithm/lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.c
#include <stdio.h>

// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    if (root == NULL)
        return NULL;
    else if (root == p || root == q)
        return root;

    struct TreeNode *left = lowestCommonAncestor(root->left, p, q),
                    *right = lowestCommonAncestor(root->right, p, q);

    if (left == NULL && right == NULL)
        return NULL;
    else if (left == NULL)
        return right;
    else if (right == NULL)
        return left;
    else
        return root;
}

int main(int argc, char* argv[])
{
    // Testing:

    return 0;
}

File: algorithm/lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.go
package lowest_common_ancestor_of_a_binary_search_tree

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	for (p.Val-root.Val)*(q.Val-root.Val) > 0 {
		if p.Val > root.Val {
			root = root.Right
		} else {
			root = root.Left
		}
	}

	return root
}
